src/trace/trace.c

```diff
@@ -96,19 +116,147 @@ static inline void mtrace_event(const char *data, uint32_t length)
 	platform_shared_commit(trace, sizeof(*trace));
 }
 
+#if CONFIG_TRACE_FILTERING_VERBOSITY
 /**
- * \brief Runtime trace filtering
+ * \brief Runtime trace filtering based on verbosity level
  * \param lvl log level (LOG_LEVEL_ ERROR, INFO, DEBUG ...)
  * \param uuid uuid address
  * \return false when trace is filtered out, otherwise true
  */
-static inline bool trace_filter_pass(uint32_t lvl,
-				     const struct tr_ctx *ctx)
+static inline bool trace_filter_verbosity(uint32_t lvl, const struct tr_ctx *ctx)
 {
 	STATIC_ASSERT(LOG_LEVEL_CRITICAL < LOG_LEVEL_VERBOSE,
 		      LOG_LEVEL_CRITICAL_MUST_HAVE_LOWEST_VALUE);
 	return lvl <= ctx->level;
 }
+#endif /* CONFIG_TRACE_FILTERING_VERBOSITY */
+
+#if CONFIG_TRACE_FILTERING_ADAPTIVE
+static void prepare_to_emit(struct recent_log_entry *entry)
+{
+	struct trace *trace = trace_get();
+	struct recent_log_entry *entry_to_emit =
+		&trace->trace_core_context[cpu_get_id()].entry_to_emit;
+
+	memcpy_s(entry_to_emit, sizeof(*entry_to_emit), entry, sizeof(*entry));
+
+	platform_shared_commit(trace, sizeof(*trace));
+
+	memset(entry, 0, sizeof(*entry));
+}
+
+static void emit_recent_entry(void)
+{
+	struct trace *trace = trace_get();
+	struct recent_log_entry *entry_to_emit =
+		&trace->trace_core_context[cpu_get_id()].entry_to_emit;
+
+	_log_message(trace_log_unfiltered, false, LOG_LEVEL_INFO, _TRACE_INV_CLASS, &dt_tr,
+		     _TRACE_INV_ID, _TRACE_INV_ID, "Suppressed %u similar messages: %pE",
+		     entry_to_emit->suppress_count, entry_to_emit->entry_id);
+
+	memset(entry_to_emit, 0, sizeof(*entry_to_emit));
+
+	platform_shared_commit(trace, sizeof(*trace));
+}
```

Comment:
to your explanation:
> @lyakh @lgirdwood The reason why I separated these functions was to maintain chronological order, to have current trace message, that's being sent, arrive before information about entries that are unsuppressed due to filter state update, caused by that message. This requires that extra step, to delay emitting entry from filter update until next message arrives, and then emit all entries that needed to be emitted up to this point.

sorry, I'm still not getting it. Every time you call `prepare_to_emit()` you know, that this entry has to be emitted. Why delay it until the next message? If the user dumps messages with a message prepared for emission and not emitted yet, they'll miss the information. And I don't see why order would be wrong. With your current approach:
- msg1
  - filter: no entries, add to recent, return true, print msg1
- msg1
  - filter: too soon, update recent, return false, msg1 suppressed
- msg1
  - filter: timeout exceeded, *prepare* "N suppressed" message, print msg1
- user dumps messages
  - msg1 printed twice, no indication of it being suppressed
- msg2
  - print "N msg1 suppressed", msg2 new, return true, print msg2

with just one print-suppressed function you'd get
- msg1
  - filter: no entries, add to recent, return true, print msg1
- msg1
  - filter: too soon, update recent, return false, msg1 suppressed
- msg1
  - filter: timeout exceeded, *print* "N suppressed" message, print msg1
- user dumps messages
  - msg1 printed, then a suppression message, then the last msg1 is printed
- msg2
  - msg2 new, return true, print msg2

Reply:
Removed `prepare` function. `emit` function now takes pointer to entry as parameter.